File: Baekjoon/Swift/DisJoint/Q1922.swift
//
//  main.swift
//  BOJ_PS
//
//  Created by 김태훈 on 2021/05/05.
//  네트워크 연결 / 최소스패닝트리

typealias Edge = (a: Int, b: Int, c: Int)

let n = Int(readLine()!)!
let m = Int(readLine()!)!

var parent = [Int](repeating: 0, count: n+1)
var edges = [Edge]()

for i in 0..<n+1 {
    parent[i] = i
}

for _ in 0..<m {
    let input = readLine()!.split(separator: " ").map { Int(String($0))! }
    edges.append((input[0], input[1], input[2]))
}

edges.sort { $0.c < $1.c }

func find(_ x: Int) -> Int {
    if parent[x] != x {
        parent[x] = find(parent[x])
    }
    return parent[x]
}

func union(_ a: Int, _ b: Int) {
    let a = find(a)
    let b = find(b)
    
    if a < b {
        parent[b] = a
    } else {
        parent[a] = b
    }
}

var result = 0
for edge in edges {
    if find(edge.a) != find(edge.b) {
        union(edge.a, edge.b)
        result += edge.c
    }
}
print(result)

File: Baekjoon/Python/BOJ_PS/DisjointSet/Q1922.py
# 네트워크 연결 / 최소 스패닝 트리
import sys

def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])
    return parent[x]

def union(a, b):
    a = find(a)
    b = find(b)

    if a < b:
        parent[b] = a
    else:
        parent[a] = b

input = sys.stdin.readline
n = int(input())
m = int(input())
parent = [i for i in range(n+1)]
edges = []

for _ in range(m):
    a, b, c = map(int, input().split())
    edges.append((a,b,c))

edges.sort(key=lambda x: x[2])

result = 0
for edge in edges:
    a, b, c = edge
    if find(a) != find(b):
        union(a, b)
        result += c
print(result)